### PR TITLE
drivers: modem: gsm_ppp: Wait for cell tower registration before attaching

### PR DESCRIPTION
**Pull request by ycsin**
`gsm_ppp` modem driver should wait for cell tower registration before the attempt to attach to the packet service.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

### CODE REVIEW

**Comment by ycsin** [comment]
On my Quectel EC21 modem, it fails to attach to the packet service when it is not registered to the cell tower yet. I'm not sure if this is a Quectel only thing.

@jukkar May I have your input please?

**Comment by ycsin** [comment]
Also, AT+CREG=2 would output extra information that would be useful to perform device locating with something like [OpenCelliD](https://opencellid.org/), however the parsing wouldn't work when the modem is not registered to the tower yet as it only output 2 of the 5 arguments which the parser expects

File: drivers/modem/gsm_ppp.c [hunk 1]
```diff
@@ -30,6 +31,15 @@ LOG_MODULE_REGISTER(modem_gsm, CONFIG_MODEM_LOG_LEVEL);
 #define GSM_RECV_BUF_SIZE               128
 #define GSM_ATTACH_RETRY_DELAY_MSEC     1000
 
+/* Modem network registration state */
+enum network_state {
+	NET_NOT_REGISTERED = 0,
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:36]: These names might conflict with other network related symbols because of the same `NET_` prefix. Could you have here `GSM_` or `GSM_NET_` instead?

File: drivers/modem/gsm_ppp.c [hunk 2]
```diff
@@ -399,6 +444,20 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
...
+				    &gsm->context.cmd_handler,
+				    &check_net_reg_cmd, 1,
+				    "AT+CREG?",
+				    &gsm->sem_response,
+				    GSM_CMD_SETUP_TIMEOUT);
+	if ((ret < 0) || ((gsm->net_state != NET_ROAMING) &&
+			 (gsm->net_state != NET_HOME_NETWORK))) {
+		k_sleep(K_SECONDS(1));
+		goto wait_reg;
+	}
```
Comment (jukkar) [inline drivers/modem/gsm_ppp.c:459]: This loop is problematic because the `gsm_finalize_connection()` is called from system workqueue and it should not block very long time. This loop can cause long sleep/delay. Because of this you need to implement this in a state machine instead.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: drivers/modem/gsm_ppp.c (hunk 2 above)
Comment (WilliamGFish) [reply drivers/modem/gsm_ppp.c:459]: Have a look at the Ublox Sara modem driver as it has already addressed this issue.

**Comment by jukkar** [comment]
@endian-benjamin you did some changes to gsm modem recently, do you have some time to review this?

File: drivers/modem/gsm_ppp.c (hunk 2 above)
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:459]: @jukkar Noted, I think I will just have to modify the [wait_reg](https://github.com/zephyrproject-rtos/zephyr/blob/70cb83d7b406e37d6bbf22eaa92c5bafa4affcad/drivers/modem/gsm_ppp.c#L447) part to reschedule the `gsm_configuration_work` in 1 seconds and simply return, then modify the [L402](https://github.com/zephyrproject-rtos/zephyr/blob/70cb83d7b406e37d6bbf22eaa92c5bafa4affcad/drivers/modem/gsm_ppp.c#L402) to jump to `wait_reg` if the `net_state` is 2 (searching). Sounds good to you?

Comment (ycsin) [reply drivers/modem/gsm_ppp.c:459]: @WilliamGFish 
> Have a look at the Ublox Sara modem driver as it has already addressed this issue.

Thanks for your input.
Previously when I was using AT+CREG=2, I tried to do it by waiting for the unsolicited command but somehow it never / took too long to happen so I switched to do AT+CREG? directly instead.

I had a look at the [sara modem](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/ublox-sara-r4.c#L1213) and they are limiting the time by a loop which worst case scenario is 41 seconds delay, I guess this is not preferred as mentioned by @jukkar

File: drivers/modem/gsm_ppp.c (hunk 1 above)
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:36]: Sure thing, I will change them to `GSM_*` tomorrow

**Comment by ycsin** [comment]
Will fix the buildkite error probably tomorrow

File: drivers/modem/gsm_ppp.c (hunk 2 above)
Comment (WilliamGFish) [reply drivers/modem/gsm_ppp.c:459]: If you look it:
Attempts to connect
If does not receive response within timeout will reset modem, then retry

If the still unconnected within the 41 secs the driver will restart (ln 1117) 
This will be repeated MDM_NETWORK_RETRY_COUNT times until then giving up and returning an error.

This has been working for a couple of years without issue.

If it can be improved that is great.

File: drivers/modem/gsm_ppp.c [hunk 3]
```diff
@@ -253,13 +267,43 @@ static const struct setup_cmd setup_cmds[] = {
 	SETUP_CMD("AT+CGSN", "", on_cmd_atcmdinfo_imei, 0U, ""),
 #endif
 
-	/* disable unsolicited network registration codes */
-	SETUP_CMD_NOHANDLE("AT+CREG=0"),
+	/* enable network registration report */
+	SETUP_CMD_NOHANDLE("AT+CREG=1"),
```
Comment (icblnk) [inline drivers/modem/gsm_ppp.c:271]: I guess there's no sense to enable NRCs as polling is done manually?

File: drivers/modem/gsm_ppp.c [hunk 4]
```diff
@@ -353,6 +400,11 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 {
 	int ret;
 
+	/* If modem is searching for network, we should skip the setup step */
+	if ((gsm->net_state == GSM_SEARCHING) && gsm->register_retries) {
+		goto registering;
+	}
```
Comment (icblnk) [inline drivers/modem/gsm_ppp.c:406]: there's no sense to check registration before attachment as attachment comes first so as for me attachment check can be replaced by the registration check and that's it.

File: drivers/modem/gsm_ppp.c (hunk 3 above)
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:271]: I think you are right, initially I enabled this to see if the modem would send the registration report periodically so that I can have a callback to my application but I think my Quectel modem doesn't really send the report unless I perform a manual query.

Will remove it when I'm back to this PR branch

File: drivers/modem/gsm_ppp.c (hunk 4 above)
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:406]: I'm not sure about this one, is it possible where a modem registered to a tower but couldn't attach to a packet service?

**Comment by ycsin** [comment]
@jukkar Modified the cell registration in `gsm_finalize_connection` sometime ago to reschedule and return if failed so that it wont block the workqvery long.

@WilliamGFish Implemented the RF reset sometime ago, it will reset if the registration timed out, @jukkar when this happens it will cause a 1 second block in the workq.

@icblnk  Remove unsolicited AT+CREG

File: drivers/modem/gsm_ppp.c [hunk 5]
```diff
@@ -35,6 +35,8 @@ LOG_MODULE_REGISTER(modem_gsm, CONFIG_MODEM_LOG_LEVEL);
 #define GSM_RECV_MAX_BUF                30
 #define GSM_RECV_BUF_SIZE               128
 #define GSM_ATTACH_RETRY_DELAY_MSEC     1000
+#define GSM_REGISTER_DELAY_MSEC         1000
+#define GSM_REGISTER_TIMEOUT            60
```
Comment (bbilas) [inline drivers/modem/gsm_ppp.c:39]: Please make it configurable via Kconfig with default value because I had cases when registration time takes about +/- 180 sec so I would like to have the possibility to override this timeout easily.

File: drivers/modem/gsm_ppp.c [hunk 6]
```diff
@@ -46,6 +48,16 @@ LOG_MODULE_REGISTER(modem_gsm, CONFIG_MODEM_LOG_LEVEL);
 	#define GSM_RSSI_MAXVAL         -51
 #endif
 
+/* Modem network registration state */
+enum network_state {
+	GSM_NOT_REGISTERED = 0,
```
Comment (bbilas) [inline drivers/modem/gsm_ppp.c:53]: By default, it is counted from 0 so it is superfluous.

**Review by jukkar — DISMISSED** [review]
LGTM

File: drivers/modem/gsm_ppp.c [hunk 7]
```diff
@@ -669,6 +719,49 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
...
+				(void)modem_cmd_send_nolock(&gsm->context.iface,
+							    &gsm->context.cmd_handler,
+							    &response_cmds[0],
+							    ARRAY_SIZE(response_cmds),
+							    "AT+CFUN=0", &gsm->sem_response,
+							    GSM_CMD_AT_TIMEOUT);
+
+				k_msleep(GSM_REGISTER_DELAY_MSEC);
+
+				(void)modem_cmd_send_nolock(&gsm->context.iface,
```
[the remark this answers is not in the document]
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c:749]: in fact I'm not 100% convinced this is the correct action if registration times out. What if the modem isn't responding to AT commands at all? We would want to power cycle the modem in that case.

To me, it would seem more natural to fail and signal this to the app somehow. But I'm not sure how, since I don't think there's a net_event for that. Perhaps it's time we did a blocking version of `gsm_ppp_start()`?

[the remark this answers is not in the document]
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:749]: > shouldn't you reset net_state here somewhere if it fails?

No need, it is set by the handler of the AT+CREG

[the remark this answers is not in the document]
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:749]: > Perhaps it's time we did a blocking version of `gsm_ppp_start()`?

Not sure about this, what I do in my code is to wait for the net connection event with a timeout, and restart the driver when timed out

[the remark this answers is not in the document]
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:749]: You are right, but it is reset in `gsm_ppp_stop`, and unsolicited output is disabled for this: https://github.com/zephyrproject-rtos/zephyr/blob/4a060c938029475140fddbe1d560dc7e08e69f33/drivers/modem/gsm_ppp.c#L476

I guess nothing except the `AT+CREG?` can set it to something else other than its reset value?

[the remark this answers is not in the document]
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c:749]: > > Perhaps it's time we did a blocking version of `gsm_ppp_start()`?
> 
> Not sure about this, what I do in my code is to wait for the net connection event with a timeout, and restart the driver when timed out

Ofc I could be wrong, but it doesn't look like you're clearing all state on all possible failures. What if net_state == SEARCHING, then we time out and reach max retries? Looks like net_state == SEARCHING will still be true

[the remark this answers is not in the document]
Comment (jukkar) [reply drivers/modem/gsm_ppp.c:749]: > Not sure about this, what I do in my code is to wait for the net connection event with a timeout, and restart the driver when timed out

It seems that we are putting lot of needed connection logic into the application. @ycsin would it be possible if you could add something to gsm_modem sample app in order to make it somewhat usable example?

[the remark this answers is not in the document]
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c:749]: But... you reschedule the work item. Once it gets executed again, won't it see the state and `goto registering`?

File: drivers/modem/gsm_ppp.c [hunk 8]
```diff
@@ -612,6 +657,11 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		goto attaching;
 	}
 
+	/* If modem is searching for network, we should skip the setup step */
+	if (gsm->net_state == GSM_SEARCHING) {
```
Comment (ycsin) [inline drivers/modem/gsm_ppp.c:661]: @endian-benjamin following your comments I do spot an issue here and I think this should be extended to include `GSM_NOT_REGISTERED`, `GSM_REGISTRATION_DENIED` & `GSM_UNKNOWN` as well

Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c:661]: Wat. That makes no sense.

Why not just reset state to NOT searching if `AT+CREG?` fails?

Comment (ycsin) [reply drivers/modem/gsm_ppp.c:661]: > Why not just reset state to NOT searching if `AT+CREG?` fails?

I don't think that we should reset it to `NOT_REGISTERED` if it is indeed searching for network, it just means that we should wait longer in this state.

I'm proposing the following:
```c
	if (gsm->net_state != GSM_ROAMING && gsm->net_state != GSM_HOME_NETWORK) {
		goto registering;
	}
```

The registration step below wants `GSM_ROAMING` / `GSM_HOME_NETWORK` before moving on to the next step. So we just make sure that we query the registration again as long as the modem is not registered at a cell tower.

Comment (ycsin) [reply drivers/modem/gsm_ppp.c:661]: That's what other modems are doing:
- [simcom](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/simcom-sim7080.c#L1154)
- [ublox sara](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/ublox-sara-r4.c#L1297)

File: drivers/modem/Kconfig.gsm [hunk 9]
```diff
@@ -79,6 +79,14 @@ config MODEM_GSM_ATTACH_TIMEOUT
 	  using AT+CGATT. This setting dictates how much time in seconds
 	  we give the modem before giving up.
 
+config MODEM_GSM_REGISTER_TIMEOUT
+	int "Timeout for registering to cellular tower"
+	default 60
```
Comment (rerickson1) [inline drivers/modem/Kconfig.gsm:84]: I think this default time should be much larger. On a modem that has never been on the network, this can take minutes. A good default for this would be something like 300 seconds.

File: drivers/modem/gsm_ppp.c [hunk 10]
```diff
@@ -669,6 +719,49 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
...
+			gsm->register_retries--;
+
+			/* Reset RF if timed out */
+			if (!gsm->register_retries) {
+				(void)modem_cmd_send_nolock(&gsm->context.iface,
+							    &gsm->context.cmd_handler,
+							    &response_cmds[0],
+							    ARRAY_SIZE(response_cmds),
+							    "AT+CFUN=0", &gsm->sem_response,
+							    GSM_CMD_AT_TIMEOUT);
```
Comment (rerickson1) [inline drivers/modem/gsm_ppp.c:745]: Is this really needed? In my experience it is better to let the modem do its own thing and let it keep searching instead of interrupting its process by turning off the radio. This should be a last resort after giving the modem minutes (~10) to try and get on a network.

**Review by rerickson1 — CHANGES_REQUESTED** [review]

**Review by WilliamGFish — CHANGES_REQUESTED** [review]
Do we need any of this additional connection code as the PDP context uses a different method ATD*99* : ~line 208 ?

This ATD performs a dial-up command which establishes the data connection. (see https://www.u-blox.com/sites/default/files/AT-CommandsExamples_AppNote_%28UBX-13001820%29.pdf)

File: drivers/modem/gsm_ppp.c (hunk 10 above)
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:745]: This was suggested by @WilliamGFish and is used in the ublox sara driver. Personally I'm not sure if this is necessary.

File: drivers/modem/gsm_ppp.c (hunk 8 above)
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c]: I'm sure you have basically the right logic in mind, but the code your proposing seems very unintuitive.

File: drivers/modem/gsm_ppp.c (hunk 10 above)
Comment (endian-benjamin) [reply drivers/modem/gsm_ppp.c:745]: I agree that it's better to just let it spin. This way the application has better control over when to give up and try again.

Comment (ycsin) [reply drivers/modem/gsm_ppp.c:745]: @rerickson1 @endian-benjamin I agree with you all

File: drivers/modem/Kconfig.gsm (hunk 9 above)
Comment (ycsin) [reply drivers/modem/Kconfig.gsm]: fixed

File: drivers/modem/gsm_ppp.c [hunk 11]
```diff
@@ -671,6 +724,30 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
...
+	ret = modem_cmd_send_nolock(&gsm->context.iface,
+				    &gsm->context.cmd_handler,
+				    &check_net_reg_cmd, 1,
+				    "AT+CREG?",
+				    &gsm->sem_response,
+				    GSM_CMD_SETUP_TIMEOUT);
+	if ((ret < 0) || ((gsm->net_state != GSM_NET_ROAMING) &&
+			 (gsm->net_state != GSM_NET_HOME_NETWORK))) {
+		if (!gsm->register_retries) {
+			gsm->register_retries = CONFIG_MODEM_GSM_REGISTER_TIMEOUT *
```
[the remark this answers is not in the document]
Comment (ycsin) [reply drivers/modem/gsm_ppp.c:735]: Kconfig is used directly a lot in this driver, I think that shouldn't be an issue? I've remove the outliers instead so make them consistent with the others.

File: drivers/modem/gsm_ppp.c [hunk 12]
```diff
@@ -617,7 +663,12 @@ static void gsm_finalize_connection(struct k_work *work)
 		goto attaching;
 	}
 
-	if (IS_ENABLED(CONFIG_GSM_MUX)) {
+	/* If modem is searching for network, we should skip the setup step */
+	if (gsm->net_state != GSM_NET_ROAMING && gsm->net_state != GSM_NET_HOME_NETWORK) {
+		goto registering;
+	}
+
+	if (IS_ENABLED(CONFIG_GSM_MUX) && gsm->mux_enabled) {
```
Comment (ycsin) [inline drivers/modem/gsm_ppp.c:671]:
```suggestion
	if (IS_ENABLED(CONFIG_GSM_MUX)) {
```

File: drivers/modem/gsm_ppp.c [hunk 13]
```diff
@@ -105,8 +118,8 @@ static struct gsm_modem {
 } gsm;
 
 NET_BUF_POOL_DEFINE(gsm_recv_pool, GSM_RECV_MAX_BUF, GSM_RECV_BUF_SIZE, 0, NULL);
-K_KERNEL_STACK_DEFINE(gsm_rx_stack, GSM_RX_STACK_SIZE);
-K_KERNEL_STACK_DEFINE(gsm_workq_stack, GSM_WORKQ_STACK_SIZE);
+K_KERNEL_STACK_DEFINE(gsm_rx_stack, CONFIG_MODEM_GSM_RX_STACK_SIZE);
```
Comment (bbilas) [inline drivers/modem/gsm_ppp.c:121]: These changes are not related to that MR at all. Please move them into separate MR.

**Review by rerickson1 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM
